### PR TITLE
Move navigation mesh baking to NavigationServer

### LINKED ISSUE #79843: NavMesh source geometry parsing ignores CSGShape3D set to `use_collision`
### Godot version

Godot 4.1

### System information

Windows 10

### Issue description

The source geometry parsing for the navigation mesh baking only adds mesh geometry from CSGShape3D when set to NavigationMesh::PARSED_GEOMETRY_MESH_INSTANCES or PARSED_GEOMETRY_BOTH.

The parsing ignores CSG shapes with `use_collision` enabled when set to PARSED_GEOMETRY_STATIC_COLLIDERS.
The parsing also ignores the `collision_layers` bitmask in all cases.

### Steps to reproduce

Add `CSGShape3D`  with `use_collision` enabled and set the NavigationMesh to `PARSED_GEOMETRY_STATIC_COLLIDERS`.

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by smix8**
Moves navigation mesh baking to NavigationServer.

- Moves the 3D `NavigationMesh` parsing and baking from `NavigationMeshGenerator` to `NavigationServer3D`.
- Deprecates the `NavigationMeshGenerator` singleton.
- Splits the monolithic geometry parsing code into node specific functions.

Fixes https://github.com/godotengine/godot/issues/79843.

Requires https://github.com/godotengine/godot/pull/79157 to be merged first as a substitute for `NavigationMeshGenerator.clear()`.

The other public functions from the NavigationMeshGenerator can already be found on the NavigationServer3D since Godot 4.1 and https://github.com/godotengine/godot/pull/77412.

Updating the NavigationMeshGenerator for further developments is not feasible as the NavigationServer needs to be in full control of the navigation mesh baking process going forward. E.g. for internal threading support or the change to a [navmesh instance system](https://github.com/godotengine/godot-proposals/issues/6317) as part of the server, or better agent sizes support. The old NavigationMeshGenerator also had nothing but problems from init chain to extendibility and required workarounds and extra functions like https://github.com/godotengine/godot/pull/79137.

### CODE REVIEW

**Comment by smix8** [comment]
Rename `NavigationGeometryParser3D` to -> `NavigationSourceGeometryParser3D` to be coherent with the `NavigationSourceGeometryData`.

The geometry parsers are now kept internal and intentionally not exposed for users. We expect that part to change when the server internals are reworked and don't want to add more compatibility issues on top the stack right now.

**Comment by smix8** [comment]
Made the node parsers more readable by splitting them into their own functions.

Now the NavigationMeshEditor uses the same function as the NavigationRegion3D for consistent bake behavior. Changed that the NavigationRegion3D did entire NavigationMesh resource duplicates for each bake which did not only drain performance on large meshes and made the behavior different in the Editor as compared to baking at runtime. It would also make the resource RID change / invalid all the time as soon as instances get enabled.

**Comment by smix8** [comment]
Added a fixed for the CSGShape3D collision and changed some functions due to changes in a subsequent pr.

File: modules/navigation/godot_navigation_server.cpp [hunk 1]
```diff
@@ -904,14 +904,37 @@ COMMAND_2(obstacle_set_avoidance_layers, RID, p_obstacle, uint32_t, p_layers) {
...
+	ERR_FAIL_COND_MSG(!p_source_geometry_data.is_valid(), "Invalid NavigationMeshSourceGeometryData3D.");
+
+	if (!p_source_geometry_data->has_data()) {
+		p_navigation_mesh->clear();
+		if (p_callback.is_valid()) {
+			Callable::CallError ce;
+			Variant result;
+			p_callback.callp(nullptr, 0, result, ce);
+			if (ce.error == Callable::CallError::CALL_OK) {
+				//
```
Comment (akien-mga) [inline modules/navigation/godot_navigation_server.cpp:929]: Is this meant to do something?

File: modules/navigation/register_types.cpp [hunk 2]
```diff
@@ -32,20 +32,19 @@
...
-#include "core/config/engine.h"
-#include "servers/navigation_server_3d.h"
-
+#ifndef DISABLE_DEPRECATED
 #ifndef _3D_DISABLED
+#include "navigation_mesh_generator.h"
+
 NavigationMeshGenerator *_nav_mesh_generator = nullptr;
 #endif
+#endif // DISABLE_DEPRECATED
```
Comment (akien-mga) [inline modules/navigation/register_types.cpp:47]: The previous include order was correct, see #75944.

It should be the "local" (`modules/navigation/` relative) includes first, then the "global" (`core/`, `servers/`, etc.) includes.

Here it makes a case which is slightly convoluted and neither convention is better than the other, but it's good to stay consistent with the rest of the codebase.

File: modules/navigation/godot_navigation_server.cpp (hunk 1 above)
Comment (smix8) [reply modules/navigation/godot_navigation_server.cpp]: leftover

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**